scripts/commands/checkPatternsIndex.ts

```diff
@@ -0,0 +1,212 @@
+  "docs/guides/execute-on-hardware.mdx",
+  "docs/guides/postprocess-results.mdx",
+];
+
+const TOC_PATH = "docs/guides/_toc.json";
+
+async function getIndexEntries(indexPath: string): Promise<string[]> {
+  const rawIndex = await readFile(indexPath, "utf-8");
+  const result: string[] = [];
+  // The index page has several unordered lists starting with *, each with links to other pages.
+  // We want to get every slug from those lists. Note that we don't care which list the slugs show
+  // up in - we only care if the slug shows up at all anywhere.
+  for (const line of rawIndex.split("\n")) {
+    if (!line.trimStart().startsWith("* ")) {
+      continue;
+    }
+
+    const slug = extractPageSlug(line);
+    if (slug) {
+      result.push(slug);
+    }
+  }
+
+  return result;
+}
+
+function extractPageSlug(text: string): string | undefined {
+  const re = /\((.*)\)/gm;
+  // Ex: '* [Circuit library](./circuit-library)'.
+  const match = re.exec(text);
+  if (!match) {
+    // Nested sections don't have any link
+    return undefined;
+  }
+  const pageSlug = match[1];
+  if (pageSlug.startsWith("http") || pageSlug.startsWith("/")) {
+    return pageSlug;
+  }
+  const page = pageSlug.split("/").pop();
+  return `/guides/${page}`;
+}
+
+function getTocSectionPageNames(sectionNode: TocEntry): string[] {
+  let results = [];
+  if (sectionNode.url) {
+    results.push(sectionNode.url);
+  }
+
+  if (sectionNode.children) {
+    for (const child of sectionNode.children) {
+      results = [...results, ...getTocSectionPageNames(child)];
+    }
+  }
+
+  return results;
+}
+
+async function getToolsTocEntriesToCheck(): Promise<string[]> {
+  const toc = JSON.parse(await readFile(TOC_PATH, "utf-8"));
+  const toolsNode = toc.children.find(
+    (child: TocEntry) => child.title == "Tools",
+  );
+  const toolsPages = getTocSectionPageNames(toolsNode);
+  return toolsPages.filter((page) => !IGNORED_URLS.includes(page));
+}
+
+async function deduplicateEntries(
+  src: string,
+  entries: string[],
+): Promise<[Set<string>, string[]]> {
+  const deduplicatedPages: Set<string> = new Set();
+  const errors: string[] = [];
+
+  for (const entry of entries) {
+    if (deduplicatedPages.has(entry)) {
+      errors.push(`❌ ${src}: The entry ${entry} is duplicated`);
+    } else {
+      deduplicatedPages.add(entry);
+    }
+  }
+
+  return [deduplicatedPages, errors];
+}
+
+function getExtraIndexPagesErrors(
+  indexPage: string,
+  indexEntries: Set<string>,
+  toolsEntries: Set<string>,
+): string[] {
+  const extraIndexPages = [...indexEntries].filter(
+    (page) => !toolsEntries.has(page),
+  );
+
+  return extraIndexPages.map(
+    (page) =>
+      `❌ ${indexPage}: The entry ${page} doesn't appear in the \`Tools\` menu.`,
+  );
+}
+
+function getExtraToolsEntriesErrors(toolsEntries: Set<string>): string[] {
+  return [...toolsEntries].map(
+    (page) => `❌ The entry ${page} is not present on any index page`,
+  );
```

Comment:
This is a little surprising to not have logic in here to do the check, and that it only creates the error messages. I see why you do it that way, but the naming should reflect it. Such as calling the arg `remainingToolsEntries`. Or, move the filtering logic inside this function.